### PR TITLE
[Feature](bangc-ops): ms_deform_attn_backward

### PR DESCRIPTION
**Pull request by Wickyzheng**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

new op me_deform_attn_backward

## 2. Modification

- docs/bangc-docs/user_guide/9_operators/index.rst
- bangc-ops/kernels/ms_deform_attn_backward/ms_deform_attn_backward.cpp
- bangc-ops/kernels/ms_deform_attn_backward/ms_deform_attn_backward_small_channels_union1.mlu
- bangc-ops/mlu_op.h
- bangc-ops/mlu_op_kernel.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/mlu_op_test.proto
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_backward/ms_deform_attn_backward.cpp
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_backward/ms_deform_attn_backward.h
- bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/ms_deform_attn_backward/test_case/case_0.prototxt
- docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [x] diff1: diff1 <= 3e-3
- [x] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          U1        |
|        3       |Layouts                               |          ARRAY      |
|        4       |Whether multi-dimensions are supported|         no                             |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           float            |
|        7       |Whether there is size limit           |             no                         |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [ ] Multi-dimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:
|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |float               |  通过        | float<br />[       OK ] ms_deform_attn_backward/TestSuite.mluOp/32 (19162 ms)<br />[----------] 33 tests from ms_deform_attn_backward/TestSuite (98205 ms total)<br />[----------] Global test environment tear-down<br />[ SUMMARY  ] Total 33 cases of 1 op(s).<br />ALL PASSED.<br />[==========] 33 test cases from 1 test suite ran. (99136 ms total)<br />[  PASSED  ] 33 test cases. [       OK ] ms_deform_attn_backward/TestSuite.mluOp/88 (2610 ms)<br />[[----------] 89 tests from ms_deform_attn_backward/TestSuite (411289 ms total)<br />[[----------] Global test<br />[environment tear-down<br />[[ SUMMARY  ] Total 89 cases of 1 op(s).<br />[ALL PASSED.<br />[==========] 89 test cases from 1 test suite ran. (412213 ms total)<br />[[  PASSED  ] 89 test cases.|
|Zero element test    |Whether to support this test      |   通过       |  仅num_levels和num_points维度支持为0元素，测试通过 |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|    通过      |   float<br />[       OK ] ms_deform_attn_backward/TestSuite.mluOp/32 (19162 ms)<br />[----------] 33 tests from ms_deform_attn_backward/TestSuite (98205 ms total)<br />[----------] Global test environment tear-down<br />[ SUMMARY  ] Total 33 cases of 1 op(s).<br />ALL PASSED.<br />[==========] 33 test cases from 1 test suite ran. (99136 ms total)<br />[  PASSED  ] 33 test cases.      |
|Mult-platform test   |MLU370/MLU590                     |    通过      |   见上述测试   |
|Nan/INF test         |Whether to support this test      |  Only support on MLU300       |     [       OK ] ms_deform_attn_backward/TestSuite.mluOp/88 (2610 ms)<br />[[----------] 89 tests from ms_deform_attn_backward/TestSuite (411289 ms total)<br />[[----------] Global test<br />[environment tear-down<br />[[ SUMMARY  ] Total 89 cases of 1 op(s).<br />[ALL PASSED.<br />[==========] 89 test cases from 1 test suite ran. (412213 ms total)<br />[[  PASSED  ] 89 test cases.    |
| 防呆测试 | - | 通过 | 1. 指针为NULL测试： PASS<br />2.shape / dim测试 ： PASS<br />3.large tensor不支持  |
| 代码覆盖率 | 测试结果| 通过|  ms_deform_attn_backward_small_channels_union1.mlu [98.2%]  98.2 % /<br />ms_deform_attn_backward_union1.mlu                [100.0%]|
| 内存泄露测试 | 测试结果| 通过|通过|

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370 x4

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|op_name| ms_deform_attn_backward   |  479536  |    81 |  0.5%  | 0.4%   |   float |  data_value: [6, 30825, 8, 32]<br />data_spatial_shapes: [4, 2]<br />data_level_start_index: [4]<br />data_sampling_loc: [6, 9664, 8, 4, 8, 2]<br />data_attn_weight: [6, 9664, 8, 4, 8]<br />grad_output: [6, 9664, 8, 32]     |
|op_name|  ms_deform_attn_backward  |  74606  |  23   |   1.3% |   0.5% |  float  | data_value: [2, 40000, 8, 32]data_spatial_shapes: [1, 2]<br />data_level_start_index: [1]<br />data_sampling_loc: [2, 40000, 8, 1, 4, 2]<br />data_attn_weight: [2, 40000, 8, 1, 4]<br />grad_output: [2, 40000, 8, 32]<br />grad_value: [2, 40000, 8, 32]    |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|  ms_deform_attn_backward  |  132000 |  51  |   0.2% | 0.8%   |  float  | data_value: [6, 30825, 8, 32]<br />data_spatial_shapes: [4, 2]<br />data_level_start_index: [4]<br />data_sampling_loc: [6, 9664, 8, 4, 8, 2]<br />data_attn_weight: [6, 9664, 8, 4, 8]<br />grad_output: [6, 9664, 8, 32]    |
|op_name|  ms_deform_attn_backward   |  25078  |   18.9 |   0.4% | 0.7%   | float  |  data_value: [2, 40000, 8, 32]data_spatial_shapes: [1, 2]<br />data_level_start_index: [1]<br />data_sampling_loc: [2, 40000, 8, 1, 4, 2]<br />data_attn_weight: [2, 40000, 8, 1, 4]<br />grad_output: [2, 40000, 8, 32]<br />grad_value: [2, 40000, 8, 32]   |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
 
+// Group:MsDeformAttn
+/*!
+ * @brief Computes the gradient of the input tensors of ::mluOpsMsDeformAttnForward.
+ *
+ * @param[in] handle
+ *   Handle to an MLUOP context that is used to manage MLU devices and queues
+ *   in the ms_deform_attn_backward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] value_desc
+ *   The descriptor of input tensor \b value tensor. For detailed information,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:5939]: The descriptor of the \b value tensor.

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[in] value
+ *   Pointer to the MLU memory that stores the input multi-scale feature maps.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:5942]: 请确认下后面的multi-scale feature maps是否正确，参数是value

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
+ * @param[in] spatial_shapes_desc
+ *   The descriptor of the \b spatial_shapes tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[in] spatial_shapes
+ *   Pointer to the MLU memory that stores the shapes of multi-scale feature maps.
+ * @param[in] level_start_index_desc
+ *   The descriptor of the \b level_start_index tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[in] level_start_index
+ *   Pointer to the MLU memory that stores the feature maps offset in data_value.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:5952]: data_value是参数吗还是？

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
+ * @par Scale Limitation
+ * - None.
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @par Note
+ * - The input \b sampling_loc contains NaN or infinity are not supported.
+ * - The \b value, \b sampling_loc, \b with attn_weight and \b grad_output  contain NaN or infinity are not
+ *   supported ON MLU500 series currently.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6013]: MLU500请确认下是否正确，之前没有出现过

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
+ * - The operation does not support MLU200 series.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:6014]: operation 》 function

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
                              const mluOpTensorDescriptor_t grad_input_desc,
                              void *grad_input);
 
+// Group:MsDeformAttn
+/*!
+ * @brief Computes the gradient of the input tensors of ::mluOpsMsDeformAttnForward.
+ *
+ * @param[in] handle
+ *   Handle to an MLUOP context that is used to manage MLU devices and queues
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:5936]: 描述需要全部和@对齐，参照wiki https://wiki.cambricon.com/pages/viewpage.action?pageId=255658083

**Review by GraceGuanH — APPROVED** [review]

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (Wickyzheng) [reply bangc-ops/mlu_op.h:6013]: 已经确认过是可以暴露出去的

**Review by GraceGuanH — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -5928,6 +5928,120 @@ mluOpMoeDispatchBackwardData(mluOpHandle_t handle,
...
+ * - The supported layout of input and output tensors must be \p MLUOP_LAYOUT_ARRAY.
+ *
+ * @par Scale Limitation
+ * - None.
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @par Note
+ * - The input \b sampling_loc that contains NaN or infinity is not supported.
```
Comment (tudejiang79) [inline bangc-ops/mlu_op.h:6011]: 这样的话，300也是不支持的，不能因为有一两个输入参数支持，就认为300支持

File: docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md [hunk 8]
```diff
@@ -0,0 +1,444 @@
+
+
+# ms_deform_attn_backward 算子开发方案设计
+* #### 文档基本信息
+
+| 项目名称    | Cambricon Neuware Neual-netwark Library                                            |
+| ----------- | ------------------------------------------------------------ |
+| 算子名称    | ms_deform_attn_backward                                                      |
+| JIRA编号    | [CNNLCORE-8313](http://jira.cambricon.com/browse/CNNLCORE-8313)|
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md:9]: CNLL去掉

File: bangc-ops/mlu_op_kernel.h [hunk 9]
```diff
@@ -546,6 +546,25 @@ void MLUOP_WIN_API mluOpUnionXKernelActiveRotatedFilterForwardHalf(
     const int kH, const int kW, const int rotations, const void *input_gdram,
     const void *indices_gdram, const void *workspace_gdram, void *output_gdram);
 
+void MLUOP_WIN_API mluOpUnion1KernelMsDeformAttnBackwardSmallChannelsKernelImpl(
```
Comment (tudejiang79) [inline bangc-ops/mlu_op_kernel.h:549]: 添加注释，参考其它算子

File: bangc-ops/kernels/ms_deform_attn_backward/ms_deform_attn_backward.cpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+#include "core/tensor.h"
+#include "core/type.h"
+#include "core/tool.h"
+#include "kernels/kernel.h"
+#include "mlu_op_kernel.h"
+
+char API[] = "[mluOpsMsDeformAttnBackward]";
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
```
Comment (tudejiang79) [inline bangc-ops/kernels/ms_deform_attn_backward/ms_deform_attn_backward.cpp:36]: 公共函数中有，调用公共函数

File: bangc-ops/mlu_op.h (hunk 7 above)
Comment (Wickyzheng) [reply bangc-ops/mlu_op.h:6011]: 这里已经指明了任何板卡的sampling_loc都不支持nan  inf.。/   在500上 value,  sampling_loc, with attn_weight and grad_output 也不支持nan  inf，在300上这些参数是支持nan inf的。  不是你说的这个意思。

File: docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+| 数据范围限制| 输入sampling_loc中不支持nan和inf。MLU500：value,sampling_loc,attn_weight and grad_output 暂时不支持NaN和inf。|
+| 原位限制    | 不支持|
+| stride限制  | 不支持|
+| 广播限制    | 不支持|
+
+### 1.5 验收标准
+
+#### 1.5.1 精度验收标准
+
+- CNNL 精度验收标准：该算子为复合类算子
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md:120]: 删除CNNL字样

File: docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+- 功能描述
+- 框架版本 + 对应源码路径
+- 需求对应网络
+- 网络中用到的规模
+- 常用规模下的竞品性能（可选）
+- 是否需要支持原位
+- 是否需要支持stride机制
+- 框架单元测试阈值指标（可选）
+- 其他特殊需求（在线量化/融合/转数提前等，可选）
+- 确认算子需求是否已经过框架层review（滤除CNNL已支持的算子）
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md:39]: 删除CNNL字样

File: docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+
+### 1.2 算子功能和应用场景描述
+
+ms_deform_attn的反向算子，计算grad_value，grad_sampling_loc和grad_attn_weight的梯度。
+
+### 1.3 算子输入输出参数要求
+
+| 参数        | 语义               | 类型（输入/输出） | 支持类型     | 物理布局 | 规模约束 |
+| ----------- | ------------------ | ----------------- | ------------ | -------- | -------- |
+| handle      | CNNL 句柄，保存运行的上下文信息。   | 输入              | mluOpHandle_t | /        | /       |
```
Comment (tudejiang79) [inline docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md:81]: 删除CNNL字样

**Review by wushaoqiang2168 — APPROVED** [review]

File: docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+                           void *grad_attn_weight);
+```
+
+## 3 实现方案设计
+
+### 3.1 实现方案
+
+ms_deform_attn_backward算子三个输出的计算方法：
+
+1、输出`grad_value`:`grad_value`的shape为[batch, num_keys, num_heads, Channels]，是输入`value`的梯度信息。在计算时通过`sampling_loc`中的坐标信息获取到`spatial_shapes`中对应的特征图尺寸`spatial_w`和`spatial_h`，在通过双线性插值得到对应点的四邻域点以及每个点对应的权重`w1`、`w2`、`w3`和`w4`，把权重乘以`grad_output`和`attn_weight`，再使用原子加将结果累加回gdram上。在mlu的实现上，可以一次处理多组`C`，若`C`大于`nram`上分配的内存，则对`C`循环处理。
```
Comment (DanieeelLiu) [inline docs/bangc-docs/design_docs/ms_deform_attn_backward/ms_deform_attn_backward.md:165]: 权重 应该替换

**Review by DanieeelLiu — APPROVED** [review]

**Review by tudejiang79 — APPROVED** [review]

**Pull request merged**